mantis-server/mantis-server-worker/src/main/java/io/mantisrx/server/worker/TaskExecutor.java

```diff
@@ -552,23 +583,114 @@ protected void updateExecutionStatus(Status status) {
     protected CompletableFuture<Void> onStop() {
         validateRunsInMainThread();
 
-        final CompletableFuture<Void> runningTaskCompletionFuture;
-        if (currentTask != null) {
-            runningTaskCompletionFuture = Services.stopAsync(currentTask, getIOExecutor());
-        } else {
-            runningTaskCompletionFuture = CompletableFuture.completedFuture(null);
+        final CompletableFuture<Void> runningTaskCompletionFuture = stopCurrentTask();
+
+        return runningTaskCompletionFuture
+            .handleAsync((dontCare, throwable) -> {
+                if (throwable != null) {
+                    log.error("Failed to stop the task successfully", throwable);
+                }
+                return stopResourceManager();
+            }, getMainThreadExecutor())
+            .thenCompose(Function.identity())
+            .whenCompleteAsync((dontCare, throwable) -> {
+                try {
+                    classLoaderHandle.close();
+                } catch (Exception e) {
+                    log.error("Failed to close classloader handle correctly", e);
+                }
```

Comment:
Yes, it does because the rationale is that the classloader handle is only within the context of the TaskExecutor, and thus it becomes the owner of the object.